File: birthday-chocolate/birthday-chocolate.js
// https://www.hackerrank.com/challenges/the-birthday-bar/

const pipe = (...fns) => initialData =>
  fns.reduce((updatedData, fn) => fn(updatedData), initialData);

const filter = fn => arr => arr.filter(fn);

const length = arr => arr.length;

const solve = (chocNums, targetSum, targetLength) =>
  pipe(
    filter(
      (_, i) =>
        chocNums
          .slice(i, i + targetLength)
          .reduce((accum, n) => accum + n, 0) === targetSum
    ),
    length
  )(chocNums);

File: birthday-chocolate/birthday-chocolate.clj
(defn birthday [choc-ns target-sum target-length]
  (let [max-i (count choc-ns)
        indexes (range max-i)]
    (->> (range (count choc-ns))
         (filter
          (fn [i]
            (let [slice-end (if (> (+ i target-length) max-i) max-i (+ i target-length))
                  slice (subvec choc-ns i slice-end)
                  sum (reduce + slice)]
              (= sum target-sum))))
         count)))